<a href="https://colab.research.google.com/github/amu1996/GCA_assignments/blob/main/FODS_Assignment_I.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Book :- 
Author :- Mark Lutz 
1. Learning Python 5th Edition O'Reilly
2.Python Cookbook  David Beazley O'Reilly 

Link:- 
1. http://www.cs.uni.edu/~fienup/cs051f10/lectures/
2. https://github.com/MrAlex6204/Books/blob/master/Learning%20Python%2C%205th%20Edition.pdf
3. https://github.com/tomarraj008/data_books/blob/master/Python%20Cookbook%2C%203rd%20Edition.pdf



First Week Assignment 

Test Your Knowledge: Quiz 

1. What are the six main reasons that people choose to use Python?
2. Name four notable companies or organizations using Python today.
3. Why might you not want to use Python in an application?
4. What can you do with Python?
5. What’s the significance of the Python import this statement?
6. Why does “spam” show up in so many Python examples in books and on the Web?
7. What is your favorite color?


Test Your Knowledge: Quiz
1. What is the Python interpreter?
2. What is source code?
3. What is byte code?
4. What is the PVM?
5. Name two or more variations on Python’s standard execution model.
6. How are CPython, Jython, and IronPython different?
7. What are Stackless and PyPy?

Test Your Knowledge: Quiz
1. How can you start an interactive interpreter session?
2. Where do you type a system command line to launch a script file?
3. Name four or more ways to run the code saved in a script file.
4. Name two pitfalls related to clicking file icons on Windows.
5. Why might you need to reload a module?
6. How do you run a script from within IDLE?
7. Name two pitfalls related to using IDLE.
8. What is a namespace, and how does it relate to module files?


Test Your Knowledge: Quiz

1. Name four of Python’s core data types.
2. Why are they called “core” data types?
3. What does “immutable” mean, and which three of Python’s core types are considered immutable?
4. What does “sequence” mean, and which three types fall into that category?
5. What does “mapping” mean, and which core type is a mapping?
6. What is “polymorphism,” and why should you care?





**Projects:-**<br>
1. Open a Python shell, enter the following expressions, and observe the results:<br>
a. 8<br>
b. 8 * 2<br>
c. 8 ** 2<br>
d. 8/12<br>
e. 8 // 12<br>
f. 8/0

In [ ]:
8
8*2
8**2
8/12
8 // 12
8/0

0.6666666666666666

a. prints the 8 and which is type int.<br>
b. * is multiplication operator(arithmetic), 8 multiplies 2 =16<br>
c. ** is power, 8 power 2 =64.<br>
d. / is divsion, 8 division by 12 = 0.6666666666666666<br>
e. // is Floor division, 8 floor division by 12 = 0<br>
f. Here 8 division by 0, which gives ZeroDivisionError: division by zero<br>


2. Write a Python program that prints (displays) your name, address, and telephone number

In [ ]:
name = 'Amit'
address = 'Imadol, Lalitpur'
number = 1234566

print(f'Name:{name}, Address:{address}, Telephone number:{number}')

Name:Amit, Address:Imadol, Lalitpur, Telephone number:1234566


Name:Amit, Address:Imadol, Lalitpur, Telephone number:1234566

3. Evaluate the following code at a shell prompt: print("Your name is", name). Then assign name an appropriate value, and evaluate the statement again.

In [ ]:
name ='XYZ'
print(f'Your name is {name}')

Your name is XYZ


Your name is XYZ

4. Write and test a program that accepts the user's name(as text) and age(as a number) as input. The program should output a sentense containing the user's name and age.

In [ ]:
name = input('Enter your name')
age = int(input('Enter your age'))
print(f"User's name {name} and age {age}")

Enter your namexyz
Enter your age34
User's name xyz and age 34


Enter your namexyz<br>
Enter your age34<br>
User's name xyz and age 34

5. Write and test a program that computes the area of a circle. This program should request a number representing  a radius as input from the user. It should use the formula 3.14*radius**2 to compute the area and then output this result suitably labeled.

In [ ]:
import math
def area(r):
  return (math.pi)*r**2
radius = int(input('enter the radius'))
print(f'Area of circle {area(radius)}')

enter the radius2
Area of circle 12.566370614359172


enter the radius2
Area of circle 12.566370614359172

6. Enter an input statement using the input function at the shll prompt. When the prompt asks you for input, enter a number. Then, attempt to add 1 to that number, observe the results and explain what happened.

In [ ]:
num = input('enter a number')
num += 1
print(num)

enter a number4


TypeError: ignored

It gives TypeError: can only concatenate str (not "int") to str because the default input of python is string  type and we can't add different data type, to add 1(int) to number input we have to map it explicitly to int type like this:<br>
num = intinput('enter a number'))<br>
num += 1<br>
print(num)<br>
This will add 1 to that number if 4 it will be 5.


7. Enter the expression help() at the shell prompt. Follow the instructions to browse the topics and modules.

In [ ]:
help()

When entered the expression help(). It calls the help function and it gets executed.
Welcome to Python 3.7's help utility!

If this is your first time using Python, you should definitely check out
the tutorial on the Internet at https://docs.python.org/3.7/tutorial/.

Enter the name of any module, keyword, or topic to get help on writing
Python programs and using Python modules.  To quit this help utility and
return to the interpreter, just type "quit".

To get a list of available modules, keywords, symbols, or topics, type
"modules", "keywords", "symbols", or "topics".  Each module also comes
with a one-line summary of what it does; to list the modules whose name
or summary contain a given string such as "spam", type "modules spam".

help>
help> topics

Here is a list of available topics.  Enter any topic name to get more help.

ASSERTION           DELETION            LOOPING             SHIFTING
ASSIGNMENT          DICTIONARIES        MAPPINGMETHODS      SLICINGS
ATTRIBUTEMETHODS    DICTIONARYLITERALS  MAPPINGS            SPECIALATTRIBUTES
ATTRIBUTES          DYNAMICFEATURES     METHODS             SPECIALIDENTIFIERS
AUGMENTEDASSIGNMENT ELLIPSIS            MODULES             SPECIALMETHODS
BASICMETHODS        EXCEPTIONS          NAMESPACES          STRINGMETHODS
BINARY              EXECUTION           NONE                STRINGS
BITWISE             EXPRESSIONS         NUMBERMETHODS       SUBSCRIPTS
BOOLEAN             FLOAT               NUMBERS             TRACEBACKS
CALLABLEMETHODS     FORMATTING          OBJECTS             TRUTHVALUE
CALLS               FRAMEOBJECTS        OPERATORS           TUPLELITERALS
CLASSES             FRAMES              PACKAGES            TUPLES
CODEOBJECTS         FUNCTIONS           POWER               TYPEOBJECTS
COMPARISON          IDENTIFIERS         PRECEDENCE          TYPES
COMPLEX             IMPORTING           PRIVATENAMES        UNARY
CONDITIONAL         INTEGER             RETURNING           UNICODE
CONTEXTMANAGERS     LISTLITERALS        SCOPING             
CONVERSIONS         LISTS               SEQUENCEMETHODS     
DEBUGGING           LITERALS            SEQUENCES           
help> ASSERTION
The "assert" statement
**********************

Assert statements are a convenient way to insert debugging assertions
into a program:

   assert_stmt ::= "assert" expression ["," expression]

The simple form, "assert expression", is equivalent to

   if __debug__:
       if not expression: raise AssertionError

The extended form, "assert expression1, expression2", is equivalent to

   if __debug__:
       if not expression1: raise AssertionError(expression2)

These equivalences assume that "__debug__" and "AssertionError" refer
to the built-in variables with those names.  In the current
implementation, the built-in variable "__debug__" is "True" under
normal circumstances, "False" when optimization is requested (command
line option "-O").  The current code generator emits no code for an
assert statement when optimization is requested at compile time.  Note
that it is unnecessary to include the source code for the expression
that failed in the error message; it will be displayed as part of the
stack trace.

Assignments to "__debug__" are illegal.  The value for the built-in
variable is determined when the interpreter starts.
help> modules

Please wait a moment while I gather a list of all available modules...

__future__          atexit              html                sched
__main__            audioop             http                scrolledlist
_abc                autocomplete        hyperparser         search
_ast                autocomplete_w      idle                searchbase
_asyncio            autoexpand          idle_test           searchengine
_bisect             base64              idlelib             secrets
_blake2             bdb                 imaplib             select
_bootlocale         binascii            imghdr              selectors
_bz2                binhex              imp                 setuptools
_cffi_backend       bisect              importlib           shelve
_codecs             browser             inspect             shlex
_codecs_cn          builtins            io                  shutil
_codecs_hk          bz2                 iomenu              sidebar
_codecs_iso2022     cProfile            ipaddress           signal
_codecs_jp          calendar            itertools           site
_codecs_kr          calltip             json                six
_codecs_tw          calltip_w           keyword             smtpd
_collections        cffi                knox                smtplib
_collections_abc    cgi                 lib2to3             sndhdr
_compat_pickle      cgitb               linecache           socket
_compression        chunk               locale              socketserver
_contextvars        cmath               logging             sqlite3
_csv                cmd                 lzma                sqlparse
_ctypes             code                macosx              squeezer
_ctypes_test        codecontext         macpath             sre_compile
_datetime           codecs              mailbox             sre_constants
_decimal            codeop              mailcap             sre_parse
_dummy_thread       collections         mainmenu            ssl
_elementtree        colorizer           marshal             stackviewer
_functools          colorsys            math                stat
_hashlib            compileall          mimetypes           statistics
_heapq              concurrent          mmap                statusbar
_imp                config              modulefinder        string
_io                 config_key          msilib              stringprep
_json               configdialog        msvcrt              struct
_locale             configparser        multicall           subprocess
_lsprof             contextlib          multiprocessing     sunau
_lzma               contextvars         mysql               symbol
_markupbase         copy                mysqlx              symtable
_md5                copyreg             netrc               sys
_msi                corsheaders         nntplib             sysconfig
_multibytecodec     crypt               nt                  tabnanny
_multiprocessing    cryptography        ntpath              tarfile
_mysql_connector    csv                 nturl2path          telnetlib
_mysqlxpb           ctypes              numbers             tempfile
_opcode             curses              opcode              test
_operator           dataclasses         operator            textview
_osx_support        datetime            optparse            textwrap
_overlapped         dbm                 os                  this
_pickle             debugger            outwin              threading
_py_abc             debugger_r          parenmatch          time
_pydecimal          debugobj            parser              timeit
_pyio               debugobj_r          pathbrowser         tkinter
_queue              decimal             pathlib             token
_random             delegator           pdb                 tokenize
_sha1               difflib             percolator          tooltip
_sha256             dis                 pickle              trace
_sha3               distutils           pickletools         traceback
_sha512             django              pip                 tracemalloc
_signal             doctest             pipes               tree
_sitebuiltins       dummy_threading     pkg_resources       tty
_socket             dynoption           pkgutil             turtle
_sqlite3            easy_install        platform            turtledemo
_sre                editor              plistlib            types
_ssl                email               poplib              typing
_stat               encodings           posixpath           undo
_string             ensurepip           pprint              unicodedata
_strptime           enum                profile             unittest
_struct             errno               pstats              urllib
_symtable           faulthandler        pty                 uu
_testbuffer         filecmp             py_compile          uuid
_testcapi           fileinput           pyclbr              venv
_testconsole        filelist            pycparser           warnings
_testimportmultiple fnmatch             pydoc               wave
_testmultiphase     format              pydoc_data          weakref
_thread             formatter           pyexpat             webbrowser
_threading_local    fractions           pyparse             window
_tkinter            ftplib              pyshell             winreg
_tracemalloc        functools           pytz                winsound
_warnings           gc                  query               wsgiref
_weakref            genericpath         queue               xdrlib
_weakrefset         getopt              quopri              xml
_winapi             getpass             random              xmlrpc
abc                 gettext             re                  xxsubtype
aifc                glob                redirector          zipapp
antigravity         grep                replace             zipfile
argparse            gzip                reprlib             zipimport
array               hashlib             rest_framework      zlib
asgiref             heapq               rlcompleter         zoomheight
ast                 help                rpc                 zzdummy
asynchat            help_about          run                 
asyncio             history             runpy               
asyncore            hmac                runscript           

Enter any module name to get more help.  Or, type "modules spam" to search
for modules whose name or summary contain the string "spam".

help> typing



8. The tax calculator program of the case study outputs a floating-point number that might show more than two digits of precision. Use the round function to modify the program to display at most two digits of precision in the output number.

In [ ]:
income = int(input('Enter your total income'))
def tax(i):
  return i*(13/100)
print('Total tax :',round(tax(income),2))

Enter your total income4444444
Total tax : 577777.72


Enter your total income4444444
Total tax : 577777.72

9. You can calculate the surface area of a cube if you know the length of an edge. Write a program that takes the length of an edge(an integer) as input and prints the cube's surface area as output.

In [ ]:
l = int(input('enter the lenght of an edge:'))
surface_area = 6 * l**2
print('The surface area of cube is', surface_area)

enter the lenght of an edge:4
The surface area of cube is 96


enter the lenght of an edge:4<br>
The surface area of cube is 96

10.Five Star Retro Video rents VHS tapes and DVDs to the same connoisseurs who like to buy LP record albums. The store rents new videos for $3.00 a night, and oldies for $2.00 a night.

Write a program that the clerks at Five Star Retro Video can use to calculate the total charge for a customer’s video rentals.

The program should prompt the user for the number of each type of video and output the total cost.

In [ ]:
new = int(input('Enter number of nights for new videos'))
oldy = int(input('Enter number of nights for old videos'))
total_price = (new * 3.00) + (oldy * 2.00)
print(f'Total cost :{total_price}')

Enter number of nights for new videos7
Enter number of nights for old videos8
Total cost :37.0


Enter number of nights for new videos7<br>
Enter number of nights for old videos8<br>
Total cost :37.0

11. Write a program that takes the radius of a sphere(a floating point number) as input and then outputs the sphere's diameter, circumference, surface area and volume.

In [ ]:
import math
r =float(input('enter the radius of sphere'))
diameter = 2 * r
circumference = 2 * (math.pi) * radius
volume = 4/3 * (math.pi) * radius ** 3
surface_area = 4 * (math.pi) * radius ** 2
print(f'Diameter of sphere:{diameter}')
print(f'Circumference of sphere:{circumference}')
print(f'Volume of sphere:{volume}')
print(f'Surface area of sphere:{surface_area}')



enter the radius of sphere5
Diameter of sphere:10.0
Circumference of sphere:12.566370614359172
Volume of sphere:33.510321638291124
Surface area of sphere:50.26548245743669


enter the radius of sphere5<br>
Diameter of sphere:10.0<br>
Circumference of sphere:12.566370614359172<br>
Volume of sphere:33.510321638291124<br>
Surface area of sphere:50.26548245743669

12. Instructions An object's momentum is its mass multiplied by its velocity. Write a program that accepts an object's mass (in kilograms) and velocity (in meters per second) as inputs, and then outputs its momentum.

In [ ]:
mass = float(input('Enter mass: '))
velocity = float(input('Enter velocity: '))
def momtm(m,v):
  return m * v
momentum = momtm(mass,velocity)
print(f'Momentum {momentum} kg*m/s')

Enter mass: 3
Enter velocity: 4
Momentum 12.0 kg*m/s


13. The kinetic energy of a moving object is given by the formula KE = ½mv2 where m is the object’s mass and v is its velocity.

Modify the program you created in Project 5 so that it prints the object’s kinetic energy as well as its momentum.

In [ ]:
mass = float(input('Enter mass:'))
velocity = float(input('Enter velocity:'))
def momtm(m,v):
  return m * v
def ke(m,v):
  return 1/2 * m * v**2
momentum = momtm(mass,velocity)
k_e = ke(mass,velocity)
print(f'Momentum is{momentum} kg*m/s')
print(f'Kinetic Energy is {k_e}')

Enter mass:5
Enter velocity:6
Momentum is30.0 kg*m/s
Kinetic Energy is 90.0


Enter mass:5<br>
Enter velocity:6<br>
Momentum is30.0 kg*m/s<br>
Kinetic Energy is 90.0

14. An employee’s total weekly pay equals the hourly wage multiplied by the total number of regular hours plus any overtime pay. Overtime pay equals the total overtime hours multiplied by 1.5 times the hourly wage. Write a program that takes as inputs the hourly wage, total regular hours, and total overtime hours and displays an employee’s total weekly pay.

In [ ]:
wage = float(input('Enter hourly wage: '))
regular_hour = float(input('Enter total regular hours: '))
overtime_hours = float(input('Enter total overtime hours: '))
def weekly_pay(w,r,o):
  return r * w + o * w * 1.5

total = weekly_pay(wage, regular_hour,overtime_hours) 

print('The employee weekly pay is', total)